### PR TITLE
React to the removal of EndpointConstraint

### PR DESCRIPTION
**Pull request by rynowak**

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ActionConstraintCache.cs [hunk 1]
```diff
@@ -165,20 +165,49 @@ private IReadOnlyList<IActionConstraint> ExtractActionConstraints(List<ActionCon
...
+                        _hasActionConstraints = found;
+                    }
+
+                    return _hasActionConstraints.Value;
+                }
+            }
         }
 
-        private struct CacheEntry
+        internal struct CacheEntry
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ActionConstraintCache.cs]: `readonly` for JAMES_POINTS++

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcEndpointDataSource.cs [hunk 2]
```diff
@@ -330,21 +329,21 @@ private bool MatchRouteValue(ActionDescriptor action, MvcEndpointInfo endpointIn
...
+                    if (actionConstraint is HttpMethodActionConstraint httpMethodActionConstraint && 
+                        !metadata.Contains(actionConstraint))
                     {
-                        metadata.Add(new HttpMethodEndpointConstraint(httpMethodActionConstraint.HttpMethods));
+                        metadata.Add(new HttpMethodMetadata(httpMethodActionConstraint.HttpMethods));
                     }
-                    else if (actionConstraint is IEndpointConstraintMetadata)
+                    else if (!metadata.Contains(actionConstraint))
                     {
-                        // The constraint might have been added earlier, e.g. it is also a filter descriptor
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcEndpointDataSource.cs]: Keep this comment?

File: test/Microsoft.AspNetCore.Mvc.Core.Test/Routing/ActionConstraintEndpointSelectorTest.cs [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+using Microsoft.AspNetCore.Mvc.Internal;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.AspNetCore.Routing.Matchers;
+using Microsoft.AspNetCore.Routing.Patterns;
+using Moq;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Mvc.Routing
+{
+    public class ActionConstraintEndpointSelectorTest
```
Comment (JamesNK) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/Routing/ActionConstraintEndpointSelectorTest.cs]: ActionConstraintMatcherPolicyTest

File: src/Microsoft.AspNetCore.Mvc.Core/ConsumesAttribute.cs [hunk 4]
```diff
@@ -192,83 +189,6 @@ public bool Accept(ActionConstraintContext context)
             return true;
         }
 
-        /// <inheritdoc />
-        public bool Accept(EndpointConstraintContext context)
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Mvc.Core/ConsumesAttribute.cs]: RIP copy paste

File: src/Microsoft.AspNetCore.Mvc.Core/Routing/ActionConstraintMatcherPolicy.cs [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+        
+        private readonly ActionConstraintCache _actionConstraintCache;
+
+        public ActionConstraintMatcherPolicy(ActionConstraintCache actionConstraintCache)
+        {
+            _actionConstraintCache = actionConstraintCache;
+        }
+
+        // Run really late.
+        public override int Order => 100000;
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Mvc.Core/Routing/ActionConstraintMatcherPolicy.cs]: So late! Is there some way to test that it is the latest?

File: src/Microsoft.AspNetCore.Mvc.Core/ConsumesAttribute.cs [hunk 6]
```diff
@@ -58,17 +58,14 @@ public ConsumesAttribute(string contentType, params string[] otherContentTypes)
...
-        /// <inheritdoc />
-        int IEndpointConstraint.Order => ConsumesActionConstraintOrder;
-
         /// <summary>
         /// Gets or sets the supported request content types. Used to select an action when there would otherwise be
         /// multiple matches.
         /// </summary>
         public MediaTypeCollection ContentTypes { get; set; }
 
+        IReadOnlyList<string> IConsumesMetadata.ContentTypes => ContentTypes;
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ConsumesAttribute.cs]: I'll take this out. This is the next PR 😆

File: test/WebSites/VersioningWebSite/VersionAttribute.cs [hunk 7]
```diff
@@ -37,10 +36,5 @@ IActionConstraint IActionConstraintFactory.CreateInstance(IServiceProvider servi
         {
             return new VersionRangeValidator(_minVersion, _maxVersion) { Order = _order ?? 0 };
         }
-
-        IEndpointConstraint IEndpointConstraintFactory.CreateInstance(IServiceProvider services)
-        {
-            return new VersionRangeValidator(_minVersion, _maxVersion) { Order = _order ?? 0 };
-        }
```
Comment (JamesNK) [inline test/WebSites/VersioningWebSite/VersionAttribute.cs]: 🎉

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ActionConstraintCache.cs (hunk 1 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Mvc.Core/Internal/ActionConstraintCache.cs]: !!!

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcEndpointDataSource.cs (hunk 2 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcEndpointDataSource.cs]: Right-o

File: src/Microsoft.AspNetCore.Mvc.Core/Routing/ActionConstraintMatcherPolicy.cs (hunk 5 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Mvc.Core/Routing/ActionConstraintMatcherPolicy.cs]: I don't really intend for it to be always always always last, I just don't want it to come before anything I can think of. There isn't too much science involed.

**Comment by rynowak** [comment]
@JamesNK any more feedback on this one?

**Comment by rynowak** [comment]
@JamesNK - please feel free to merge your MVC CORS PR. I'll rebase on top later tonite

**Pull request merged**